sql: adjust DistSQL physical planning in multi-tenant mode

yuzefovich opened the pull request:
This commit removes the logic that might reassign some spans to the
gateway that can be applied in the multi-tenant deployment mode. This
logic was introduced in https://github.com/cockroachdb/cockroach/pull/80353 with the following [rationale](https://github.com/cockroachdb/cockroach/pull/80353#pullrequestreview-961275076):
```
This will probably go away once we support locality-aware distribution,
at least in this form. There's two reasons why I did this right now:
1. It avoids an extra hop (KV->assignee->gateway->client vs
KV->gateway->client), and this is an optimization we sometimes do in
the non-MT code path though at a later stage,
2. It makes the assignments and distribution type deterministic in
testing when we expect to assign to only a single pod.
```
Since then the locality-aware planning has been implemented (addressing
first half of point 1.). Also, the second half of point 1. applies both
to single-tenant and multi-tenant modes (`maybeMoveSingleFlowToGateway`).
Point 2. is a bit unfortunate and I'm not sure what to do about it yet
(we'll need to figure it out if we ever make separate-process
multi-tenant the default mode for running tests that check DistSQL
planning). For now only a single test needed an adjustment to make it
deterministic.

Additionally, running TPCH queries experimentally has shown that this
reassigning of single TableReader can make the query latency
significantly different (sometimes reduce it, sometimes increase it)
(see [here](https://github.com/cockroachdb/cockroach/issues/104379#issuecomment-1599208200)).

All of these reasons suggest that we should just get rid off this logic
to unify the DistSQL physical planner more between single-tenant and
multi-tenant modes.

Addresses: #104379.
Epic: CRDB-26687

Release note: None

yuzefovich committed:
sql: adjust DistSQL physical planning in multi-tenant mode

This commit removes the logic that might reassign some spans to the
gateway that can be applied in the multi-tenant deployment mode. This
logic was introduced in #80353 with the following [rationale](https://github.com/cockroachdb/cockroach/pull/80353#pullrequestreview-961275076):
```
This will probably go away once we support locality-aware distribution,
at least in this form. There's two reasons why I did this right now:
1. It avoids an extra hop (KV->assignee->gateway->client vs
KV->gateway->client), and this is an optimization we sometimes do in
the non-MT code path though at a later stage,
2. It makes the assignments and distribution type deterministic in
testing when we expect to assign to only a single pod.
```
Since then the locality-aware planning has been implemented (addressing
first half of point 1.). Also, the second half of point 1. applies both
to single-tenant and multi-tenant modes (`maybeMoveSingleFlowToGateway`).
Point 2. is a bit unfortunate and I'm not sure what to do about it yet
(we'll need to figure it out if we ever make separate-process
multi-tenant the default mode for running tests that check DistSQL
planning). For now only a single test needed an adjustment to make it
deterministic.

Additionally, running TPCH queries experimentally has shown that this
reassigning of single TableReader can make the query latency
significantly different (sometimes reduce it, sometimes increase it)
(see [here](https://github.com/cockroachdb/cockroach/issues/104379#issuecomment-1599208200)).

All of these reasons suggest that we should just get rid off this logic
to unify the DistSQL physical planner more between single-tenant and
multi-tenant modes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105256)
<!-- Reviewable:end -->

knz commented:
> All of these reasons suggest that we should just get rid off this logic to unify the DistSQL physical planner more between single-tenant and multi-tenant modes.

I agree with unification as a starting point. If we want to implement the extra optimization, this should be done 1) in a way independent of multi-tenancy and 2) with an eye towards experimental verification e.g. with TPC-H.

Thanks for this PR.

yuzefovich committed:
sqlinstance: remove unused error

The only usages of this error were removed in ea7fffb138e44b28f9c37fb02a7d13df7fec7782.

Epic: None

Release note: None

yuzefovich commented:
Now that #105451 merged, the CI here should be green, RFAL.

knz reviewed (approved):
> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105256)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rharding6373)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
